File: src/TCG/Translator.cpp
#include "Common/Common.h"
#include "TCG/Translator.h"

Translator::Translator(std::shared_ptr<TACFile> TACFile_, std::shared_ptr<Scope> Global_Scpoe_) : TACFile_(TACFile_), Global_Scope(Global_Scpoe_) {
     // todo 关于如何输入TAC待定
     ASMFile_ = std::shared_ptr<ASMFile>(new ASMFile);
}


void Translator::Translate() {
    this->dataTranslate();
    this->textTranslate();
}

void Translator::dataTranslate() {
    ASMSection ASMSection_;
    ASMSection_.name = ".data";

    // todo 根据符号表的静态, SymbolManager_(Scope_) 变量，生成数据段
    // todo 需要提供global的scope
    // todo 在global的scope里面找全局变量
    ASMBlock ASMBlock_;
    ASMBlock_.name = "";
    std::shared_ptr<TACBlock> global = (*TACFile_).at("global");
    for (int i = 0; i < global->size(); i++) {
        std::string ASMLine_;
        switch ((*global)[i].op)
        {
            case TACOP::ASSIGN: {
                ASMLine_ = (*global)[i].dst.value + "   equ " + (*global)[i].src1.value;
                ASMBlock_.asmlines.push_back(ASMLine_);
                break;
            }
            default: {
                LOG(ERROR) << "global op error: " << to_string((*global)[i].op);
                break;
            }
        }

    }

    ASMSection_.asmblocks.push_back(ASMBlock_);
    ASMFile_->push_back(ASMSection_);
}

void Translator::textTranslate() {
    ASMSection ASMSection_;
    ASMSection_.name = ".text";

    // todo 加入global start语句
    ASMBlock ASMBlock_;
    ASMBlock_.name = "";
    ASMBlock_.asmlines.push_back("global _start");
    ASMBlock_.asmlines.push_back("extern myprint");
    ASMSection_.asmblocks.push_back(ASMBlock_);

    // todo 如何遍历TACblock待定
    // todo 以函数为单位
    for(auto i = TACFile_->begin(); i != TACFile_->end(); i++) {
        /* crTODO: 将 SymbolManager_ 改为 一个快一个 ? ljh 不用 */
        // todo 根据函数名到block的map初始化
        if (i->first == "global" || i->second->size() == 0) continue;
        SymbolManager SymbolManager_(Global_Scope, i->first);
        ASMBlock ASMBlock_ = BlockTranslator_.BlockTranslate(SymbolManager_, i->second);
        ASMSection_.asmblocks.push_back(ASMBlock_);
    }

    ASMFile_->push_back(ASMSection_);
}

void Translator::OutputFile(std::string filename) {
    std::ofstream outfile;
    outfile.open(filename);


    for (int i = 0; i < ASMFile_->size(); i++) {
        if ((*ASMFile_)[i].name != "") {
            outfile << (*ASMFile_)[i].name << ":" << std::endl;
        }
        for (int j = 0; j < (*ASMFile_)[i].asmblocks.size(); j++) {
            if ((*ASMFile_)[i].asmblocks[j].name != "") {
                outfile << (*ASMFile_)[i].asmblocks[j].name << ":" << std::endl;
            }
            for (int k = 0; k < (*ASMFile_)[i].asmblocks[j].asmlines.size(); k++) {
                outfile << (*ASMFile_)[i].asmblocks[j].asmlines[k] << std::endl;
            }
            outfile << std::endl;
        }
        outfile << std::endl;
    }

    outfile.close();
}

File: go_compiler/tcg/Translator.py
import pickle
from loguru import logger
from ..tcg.SymbolManager import SymbolManager
from ..common.tac import TACOP, TACBlock, TACFile
from ..common.Scope import Scope
from ..tcg.BlockTranslator import ASMBlock, BlockTranslator
from ..tcg.asm import ASMFile, ASMSection


class Translator:
    def __init__(self, TACFile_: TACFile, Global_Scope_: Scope | None):
        self.BlockTranslator_: BlockTranslator = BlockTranslator()
        self.TACFile_ = TACFile_
        self.Global_Scope = Global_Scope_
        self.ASMFile_: ASMFile = []

    # BlockTranslator BlockTranslator_;
    # std::shared_ptr<Scope> Global_Scope;
    # std::shared_ptr<TACFile> TACFile_;
    # std::shared_ptr<ASMFile> ASMFile_;

    def dataTranslate(self):
        ASMSection_ = ASMSection()
        ASMSection_.name = ".data"

        # TODO: 根据符号表的静态, SymbolManager_(Scope_) 变量，生成数据段
        # TODO: 需要提供global的scope
        # TODO: 在global的scope里面找全局变量
        ASMBlock_ = ASMBlock()
        ASMBlock_.name = ""
        global_: TACBlock = self.TACFile_["global"]
        # logger.debug(f"TACFile: {self.TACFile_}")
        for i in range(len(global_)):
            ASMLine_: str = ""
            if global_[i].op == TACOP.ASSIGN:
                ASMLine_ = global_[i].dst.value + "   equ  " + global_[i].src1.value
                ASMBlock_.asmlines.append(ASMLine_)
            else:
                logger.error("global op error: " + f"{pickle.dumps(global_[i].op)}")

        ASMSection_.asmblocks.append(ASMBlock_)
        self.ASMFile_.append(ASMSection_)

    def textTranslate(self):
        ASMSection_ = ASMSection()
        ASMSection_.name = ".text"

        # todo 加入global start语句
        ASMBlock_ = ASMBlock()
        ASMBlock_.name = ""
        ASMBlock_.asmlines.append("global _start")
        ASMBlock_.asmlines.append("extern myprint")
        ASMSection_.asmblocks.append(ASMBlock_)

        # todo 如何遍历TACblock待定
        # todo 以函数为单位
        for key, value in self.TACFile_.items():
            # TODO: 根据函数名到block的map初始化
            if key == "global" or key == "myprint" or len(value) == 0:
                continue
            logger.debug("start 翻译函数: " + key)
            SymbolManager_: SymbolManager = SymbolManager(self.Global_Scope, key)
            # logger.debug(f"InputValue: {key} : {value}")
            ASMBlock_: ASMBlock = self.BlockTranslator_.BlockTranslate(
                SymbolManager_, value
            )

            ASMSection_.asmblocks.append(ASMBlock_)

        self.ASMFile_.append(ASMSection_)

    def Translate(self):
        self.dataTranslate()
        self.textTranslate()

    def OutputFile(self, filename: str):
        logger.debug(f"Writing {filename}")
        with open(filename, "w") as outfile:
            for i in range(len(self.ASMFile_)):
                if self.ASMFile_[i].name != "":
                    outfile.write(self.ASMFile_[i].name + ":\n")
                for j in range(len(self.ASMFile_[i].asmblocks)):
                    if self.ASMFile_[i].asmblocks[j].name != "":
                        outfile.write(self.ASMFile_[i].asmblocks[j].name + ":\n")
                    for k in range(len(self.ASMFile_[i].asmblocks[j].asmlines)):
                        outfile.write(self.ASMFile_[i].asmblocks[j].asmlines[k] + "\n")
                    outfile.write("\n")
                outfile.write("\n")
